networking/kvm.go

```diff
@@ -0,0 +1,231 @@
+	if err != nil {
+		return nil, fmt.Errorf("tuntap persist %v", err)
+	}
+	link, err := netlink.LinkByName(ifName)
+	if err != nil {
+		return nil, fmt.Errorf("cannot find link %q: %v", ifName, err)
+	}
+	err = netlink.LinkSetUp(link)
+	if err != nil {
+		return nil, fmt.Errorf("cannot set link up %q: %v", ifName, err)
+	}
+	return link, nil
+}
+
+// kvmSetupNetAddressing calls IPAM plugin (with a hack) to reserve an IP to be
+// used by newly create tuntap pair
+// in result it updates activeNet.runtime configuration with IP, Mask and HostIP
+func kvmSetupNetAddressing(network *Networking, n activeNet, ifName string) error {
+	// TODO: very ugly hack, that go through upper plugin, down to ipam plugin
+	if err := ip.EnableIP4Forward(); err != nil {
+		return fmt.Errorf("failed to enable forwarding: %v", err)
+	}
+	n.conf.Type = n.conf.IPAM.Type
+	output, err := network.execNetPlugin("ADD", &n, ifName)
+	if err != nil {
+		return fmt.Errorf("problem executing network plugin %q (%q): %v", n.conf.Type, ifName, err)
+	}
+
+	result := plugin.Result{}
+	if err = json.Unmarshal(output, &result); err != nil {
+		return fmt.Errorf("error parsing %q result: %v", n.conf.Name, err)
+	}
+
+	if result.IP4 == nil {
+		return fmt.Errorf("net-plugin returned no IPv4 configuration")
+	}
+
+	n.runtime.IP, n.runtime.Mask, n.runtime.HostIP = result.IP4.IP.IP, net.IP(result.IP4.IP.Mask), result.IP4.Gateway
+	return nil
+}
+
+// kvmSetup prepare new Networking to be used in kvm environment based on tuntap pair interfaces
+// to allow communication with virtual machine created by lkvm tool
+// right now it only supports default "ptp" network type (other types ends with error)
+func kvmSetup(podRoot string, podID types.UUID, fps []ForwardedPort, privateNetList common.PrivateNetList, localConfig string) (*Networking, error) {
+	network := Networking{
+		podEnv: podEnv{
+			podRoot:      podRoot,
+			podID:        podID,
+			netsLoadList: privateNetList,
+			localConfig:  localConfig,
+		},
+	}
+	var e error
+	network.nets, e = network.loadNets()
+	if e != nil {
+		return nil, fmt.Errorf("error loading network definitions: %v", e)
+	}
+
+	for _, n := range network.nets {
+		switch n.conf.Type {
+		case "ptp":
+			link, err := setupTapDevice()
+			if err != nil {
+				return nil, err
+			}
+			ifName := link.Attrs().Name
+			n.runtime.IfName = ifName
+
+			err = kvmSetupNetAddressing(&network, n, ifName)
+			if err != nil {
+				return nil, err
+			}
+
+			// add address to host tap device
+			err = netlink.AddrAdd(
+				link,
+				&netlink.Addr{
+					IPNet: &net.IPNet{
+						IP:   n.runtime.HostIP,
+						Mask: net.IPMask(n.runtime.Mask),
+					},
+					Label: ifName,
+				})
+			if err != nil {
+				return nil, fmt.Errorf("cannot add address to host tap device %q: %v", ifName, err)
+			}
+
+			if n.conf.IPMasq {
+				h := sha512.Sum512([]byte(podID.String()))
+				chain := fmt.Sprintf("CNI-%s-%x", n.conf.Name, h[:8])
```

Comment:
It would be nice to factor the chain name generation in a function since the logic is duplicated both here and in `teardownKvmNets`.

But I suggest to do that after this PR gets merged since there is the same thing in CNI.